### PR TITLE
[SPARK-17190] [SQL] Removal of HiveSharedState

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

Since `HiveClient` is used to interact with the Hive metastore, it should be hidden in `HiveExternalCatalog`. After moving `HiveClient` into `HiveExternalCatalog`, `HiveSharedState` becomes a wrapper of `HiveExternalCatalog`. Thus, removal of `HiveSharedState` becomes straightforward. After removal of `HiveSharedState`, the reflection logic is directly applied on the choice of `ExternalCatalog` types, based on the configuration of `CATALOG_IMPLEMENTATION`. 

~~`HiveClient` is also used/invoked by the other entities besides HiveExternalCatalog, we defines the following two APIs: getClient and getNewClient~~
### How was this patch tested?

The existing test cases

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @rxin @yhuai @liancheng @clockfly

**Comment by cloud-fan** [comment]
is `getNewClient` equal to `getClient.newSession`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -41,13 +42,20 @@ import org.apache.spark.sql.types.{DataType, StructType}
...
   extends ExternalCatalog with Logging {
 
   import CatalogTypes.TablePartitionSpec
   import HiveExternalCatalog._
   import CatalogTableType._
 
+  /**
+   * A Hive client used to interact with the metastore.
+   */
+  private val client: HiveClient = {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: can we make it public and remove `getClient`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala [hunk 2]
```diff
@@ -40,6 +40,15 @@ abstract class ExternalCatalogSuite extends SparkFunSuite with BeforeAndAfterEac
 
   protected def resetState(): Unit = { }
 
+  // Clear all state before each test
+  override def beforeEach(): Unit = {
+    try {
+      resetState()
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: why do we need this now?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 3]
```diff
@@ -139,12 +139,9 @@ private[hive] class TestHiveSparkSession(
 
   assume(sc.conf.get(CATALOG_IMPLEMENTATION) == "hive")
 
-  // TODO: Let's remove HiveSharedState and TestHiveSessionState. Otherwise,
-  // we are not really testing the reflection logic based on the setting of
-  // CATALOG_IMPLEMENTATION.
   @transient
-  override lazy val sharedState: HiveSharedState = {
-    existingSharedState.getOrElse(new HiveSharedState(sc))
+  override lazy val sharedState: SharedState = {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: do we need to override this?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 4]
```diff
@@ -139,12 +139,9 @@ private[hive] class TestHiveSparkSession(
 
   assume(sc.conf.get(CATALOG_IMPLEMENTATION) == "hive")
 
-  // TODO: Let's remove HiveSharedState and TestHiveSessionState. Otherwise,
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: is `TestHiveSessionState` already removed?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Sure, will do it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: https://github.com/apache/spark/blob/7bb64aae27f670531699f59d3f410e38866609b7/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala#L31-L35

Before this PR, we create a metastore at a temp location to avoid any potential conflict of having multiple connections to a single derby instance. This does not sound an issue now. Is my understanding right?

If we do not create a new metastore, we have to clean the existing metastore; otherwise, [the checking](https://github.com/apache/spark/blob/5effc016c893ce917d535cc1b5026d8e4c846721/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala#L60-L61) in the test case could fail.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: Yeah. Let me move it back.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: `sharedState` in `SparkSession` is unable to access the `existingSharedState` defined in `TestHiveSparkSession`.  We are unable to override it because [it is private](https://github.com/apache/spark/blob/18c2c92580bdc27aa5129d9e7abda418a3633ea6/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L55).

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: The comment is moved to https://github.com/gatorsmile/spark/blob/d61ced8a2e81cef13afa5e0aa6203e505dd67570/sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala#L147-L148

Thanks!

**Comment by gatorsmile** [comment]
@cloud-fan Yeah, `getNewClient` equals to `getClient.newSession`. After rethinking it, maybe we do not need it. : ) Thus, the latest changes remove it too.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: But is `existingSharedState.getOrElse(new SharedState(sc))` same with what the parent do?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: Although `existingSharedState` is using the same name, we did not override [the one](https://github.com/apache/spark/blob/18c2c92580bdc27aa5129d9e7abda418a3633ea6/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L55) in `SharedState`. Thus, they are different. 

Actually, I tried it. It breaks [one test case](https://github.com/apache/spark/blob/a117afa7c2d94f943106542ec53d74ba2b5f1058/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala#L1082) 

BTW, it works if `TestHiveSparkSession` overrides both [`sparkContext` and `existingSharedState`](https://github.com/apache/spark/blob/18c2c92580bdc27aa5129d9e7abda418a3633ea6/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L54-L55). However, we have to change `existingSharedState` in `SparkSession` to non-private.

Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: i see, thanks

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: Then can we remove the `afterEach` now?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDataFrameSuite.scala [hunk 5]
```diff
@@ -31,7 +31,7 @@ class HiveDataFrameSuite extends QueryTest with TestHiveSingleton {
   }
 
   test("SPARK-15887: hive-site.xml should be loaded") {
-    val hiveClient = spark.sharedState.asInstanceOf[HiveSharedState].metadataHive
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDataFrameSuite.scala]: do we need to change this line?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSparkSubmitSuite.scala [hunk 6]
```diff
@@ -378,10 +378,9 @@ object SetMetastoreURLTest extends Logging {
         s"spark.sql.test.expectedMetastoreURL should be set.")
     }
 
-    // HiveSharedState is used when Hive support is enabled.
+    // HiveExternalCatalog is used when Hive support is enabled.
     val actualMetastoreURL =
-      spark.sharedState.asInstanceOf[HiveSharedState]
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSparkSubmitSuite.scala]: do we need to change this line?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 7]
```diff
@@ -51,7 +51,8 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
 
   // To test `HiveExternalCatalog`, we need to read the raw table metadata(schema, partition
   // columns and bucket specification are still in table properties) from hive client.
-  private def hiveClient: HiveClient = sharedState.asInstanceOf[HiveSharedState].metadataHive
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: do we need to change this line?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala [hunk 8]
```diff
@@ -266,7 +266,7 @@ class ShowCreateTableSuite extends QueryTest with SQLTestUtils with TestHiveSing
   }
 
   private def createRawHiveTable(ddl: String): Unit = {
-    hiveContext.sharedState.asInstanceOf[HiveSharedState].metadataHive.runSqlHive(ddl)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: do we need to change this line?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: Since the test cases in `ExternalCatalog` do not drop the created tables, we need to drop it. To avoid duplicate cleanup, we can change `afterEach` to `afterAll`. Let me fix it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDataFrameSuite.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDataFrameSuite.scala]: `HiveSharedState` is removed, and thus, we have to do it. Right?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSparkSubmitSuite.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSparkSubmitSuite.scala]: See the above answer.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: See the above answer.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: See the above answer.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDataFrameSuite.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDataFrameSuite.scala]: oh sorry I misread the code, `metadataHive` only exist in `HiveContext` now.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala [hunk 9]
```diff
@@ -40,6 +40,15 @@ abstract class ExternalCatalogSuite extends SparkFunSuite with BeforeAndAfterEac
 
   protected def resetState(): Unit = { }
 
+  // Clear the state before all tests
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: Does that mean the newly created hive client already has data in it?

Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: Yeah, it connects to the same metastore. We can change it by setting `HiveConf.ConfVars.METASTORECONNECTURLKEY.varname` when creating a new external catalog.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala [hunk 10]
```diff
@@ -21,26 +21,26 @@ import org.apache.hadoop.conf.Configuration
 
 import org.apache.spark.SparkConf
 import org.apache.spark.sql.catalyst.catalog._
-import org.apache.spark.sql.hive.client.HiveClient
 
 /**
  * Test suite for the [[HiveExternalCatalog]].
  */
 class HiveExternalCatalogSuite extends ExternalCatalogSuite {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: Why the changes in this file make hive client connect to a metastore that already has data?

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: Before the PR, what `HiveExternalCatalogSuite` uses is a [HiveUtils.newClientForExecution](https://github.com/apache/spark/blob/7bb64aae27f670531699f59d3f410e38866609b7/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala#L34) . The `newClientForExecution`'s configuration is `newTemporaryConfiguration`, [which makes a new path for metastore](https://github.com/apache/spark/blob/2ae7b88a07140e012b6c60db3c4a2a8ca360c684/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala#L366-L379). Thus, we can say it is pointing to a different metastore.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala [hunk 11]
```diff
@@ -21,26 +21,26 @@ import org.apache.hadoop.conf.Configuration
...
+  private val externalCatalog: HiveExternalCatalog = {
+    new HiveExternalCatalog(new SparkConf, new Configuration)
   }
 
   protected override val utils: CatalogTestUtils = new CatalogTestUtils {
     override val tableInputFormat: String = "org.apache.hadoop.mapred.SequenceFileInputFormat"
     override val tableOutputFormat: String = "org.apache.hadoop.mapred.SequenceFileOutputFormat"
-    override def newEmptyCatalog(): ExternalCatalog =
-      new HiveExternalCatalog(client, new Configuration())
+    // We create a metastore at a temp location to avoid any potential
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: is it still valid?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala [hunk 12]
```diff
@@ -21,26 +21,26 @@ import org.apache.hadoop.conf.Configuration
...
  * Test suite for the [[HiveExternalCatalog]].
  */
 class HiveExternalCatalogSuite extends ExternalCatalogSuite {
 
-  private val client: HiveClient = {
-    // We create a metastore at a temp location to avoid any potential
-    // conflict of having multiple connections to a single derby instance.
-    HiveUtils.newClientForExecution(new SparkConf, new Configuration)
+  private val externalCatalog: HiveExternalCatalog = {
+    new HiveExternalCatalog(new SparkConf, new Configuration)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: how about
```
val catalog = ....
catalog.client.reset()
catalog
```

then we don't need https://github.com/apache/spark/pull/14757/files#diff-8c4108666a6639034f0ddbfa075bcb37R43, which is too far away from here and hard to connect them together.

**Comment by cloud-fan** [comment]
LGTM except the last 2 comments

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala (hunk 11 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: It is not valid now. Will remove it. Thanks!

**Comment by gatorsmile** [comment]
Thank you!

**Comment by cloud-fan** [comment]
thanks. merging to master!
